libcxx/test/std/algorithms/alg.nonmodifying/alg.contains/ranges.contains.pass.cpp

```diff
@@ -0,0 +1,299 @@
+    int a[]              = {1, 9, 0, 13, 25};
+    int projection_count = 0;
+    {
+      bool ret = std::ranges::contains(a, a + 5, 0, [&](int i) {
+        ++projection_count;
+        return i;
+      });
+      assert(ret);
+      assert(projection_count == 3);
+      projection_count = 0;
+    }
+    {
+      bool ret = std::ranges::contains(a, 0, [&](int i) {
+        ++projection_count;
+        return i;
+      });
+      assert(ret);
+      assert(projection_count == 3);
+    }
+  }
+
+  { // check invocations of the projection for std::string
+    const std::string str{"hello world"};
+    const std::string str1{"hi world"};
+    int projection_count = 0;
+    {
+      std::string a[] = {str1, str1, str, str1, str1};
+      auto whole =
+        std::ranges::subrange(forward_iterator(std::move_iterator(a)), forward_iterator(std::move_iterator(a + 5)));
+      bool ret = std::ranges::contains(whole.begin(), whole.end(), "hello world", [&](const std::string i) {
+        ++projection_count;
+        return i;
+      });
+      assert(ret);
+      assert(projection_count == 3);
+      projection_count = 0;
+    }
+    {
+      std::string a[] = {str1, str1, str, str1, str1};
+      auto whole =
+        std::ranges::subrange(forward_iterator(std::move_iterator(a)), forward_iterator(std::move_iterator(a + 5)));
+      bool ret = std::ranges::contains(whole, "hello world", [&](const std::string i) {
+        ++projection_count;
+        return i;
+      });
+      assert(ret);
+      assert(projection_count == 3);
+    }
+  }
+
+  { // check invocations of the projection for non-continuous iterators
+    std::vector<bool> whole{false, false, true, false};
+    int projection_count = 0;
+    {
+      bool ret = std::ranges::contains(whole.begin(), whole.end(), true, [&](int i) {
+        ++projection_count;
+        return i;
+      });
+      assert(ret);
+      assert(projection_count == 3);
+      projection_count = 0;
+    }
+    {
+      bool ret = std::ranges::contains(whole, true, [&](int i) {
+        ++projection_count;
+        return i;
+      });
+      assert(ret);
+      assert(projection_count == 3);
+    }
+  }
+
+  { // check invocations of the projection for views::transform
+    int a[]              = {1, 2, 3, 4, 5};
+    int projection_count = 0;
+    auto square_number   = a | std::views::transform([](int x) { return x * x; });
+    {
+      bool ret = std::ranges::contains(square_number.begin(), square_number.end(), 16, [&](int i) {
+        ++projection_count;
+        return i;
+      });
+      assert(ret);
+      assert(projection_count == 4);
+      projection_count = 0;
+    }
+    {
+      bool ret = std::ranges::contains(square_number, 16, [&](int i) {
+        ++projection_count;
+        return i;
+      });
+      assert(ret);
+      assert(projection_count == 4);
+    }
+  }
+
+  return true;
+}
+
+// count invocations of the projection for std::list
+bool test_nonconstexpr() {
```

Comment:
Why do we test with `std::list` specifically? AFAIK we don't do that anywhere else.

Reply:
Because it doesn't provide constexpr contiguous iterators. We want to test against types that not only model different concepts, but fundamentally are different things at runtime too. 

Maybe the comment should be updated with the rational? Or removed entirely.

Reply:
How do we test anything differently w.r.t. `constexpr`-ness of the iterators? If they are passed as runtime objects the library can't do anything with them at compile time. And FWIW `std::list` will inevitably become `constexpr` too. I think a comment and maybe a bug report that we should test with different data structures would be good too.

Reply:
> We want to test against types that not only model different concepts, but fundamentally are different things at runtime too.

What kind of issues do you expect to find? I'd also prefer to do this kind of change en masse; this algorithm now differs from the general testing pattern for no inherent reason.

Reply:
>  I'd also prefer to do this kind of change en masse; this algorithm now differs from the general testing pattern for no inherent reason.

Please don't gate improvements on existing bad practice.

If you would like to make this improvement en mass, I would be happy to review it.

> What kind of issues do you expect to find? 

I expect tests like this to catch undefined behavior by changing codegen significantly which can enable

(1) Sanitizers to catch it. 
(2) Optimizations to expose it.

Moving from a contiguous array of elements to something that isn't a contiguous array of elements is an obvious variation. 

Other ways to get diversity to tickle out this sort of bug include:

*  Testing with higher alignment types.
* Testing types with different calling conventions.
 
These aren't compile time tests, you actually have to run them for them to be of value.

Reply:
> Please don't gate improvements on existing bad practice.

I don't think we should be doing this. There is, however, a certain value in consistency as well. I'm fine with a test being an outlier, whether temporarily so or not, as long as the improvement is worth the inconsistency.

> I expect tests like this to catch undefined behavior by changing codegen significantly which can enable

Can you give a more specific example? This makes it sound like testing for incorrect behavior by the compiler.

Reply:
I support adding tests for non-contiguous containers. We should add them to the other algorithms' tests in early 2024.

Reply:
FWIW, I just saw this but I would support testing all of our algorithms against all of our containers. But I would expect such testing to happen via something like a `robust_against_foo.pass.cpp` test, not in the tests for each individual algorithm. As it stands, this test is slightly inconsistent with what we do elsewhere, and I don't think it's fair to say that what we do everywhere is "bad practice".

Anyway, I don't think it's a big deal and it's not worth changing this now, but in the future I think we should do these kinds of tests as to stay consistent with the rest of our test suite, and we should consider doing it en-masse.